### ISSUE TITLE
Horizontal glitches while scrolling on AMD Radeaon Graphics

### ISSUE THREAD

**Issue by mHahnle** [canonical/mir#3527]
**Setup:**
Ubuntu Server 22.04.04 + Ubuntu Frame + Wpe Webkit Mir Kiosk

**Symptoms:**
When scrolling, horizontal glitches with random vertical positions flicker on the screen

**Tested Hardware:**
- AMD Ryzen 5 5600G (internal Radeon GPU) with different mainboards (also different CPUs of the same kind):
- Intel Core i3-8100T (internal Intel UHD Graphics 630)
- Intel Core i5-7600T (internal Intel UHD Graphics 630)
- VMWare Workstation VM (on Host with AMD Ryzen 9 7950x3d and NVidia RTX 4070 Super)

**Tested Software/OS:**
- Ubuntu Server 22.04.04
- Ubuntu Desktop 22.04.04
- Ubuntu Frame (latest version in snap on channel 22)
- WPE Webkit Mir Kiosk (latest version in snap)

**Result:**
No problems at all on both Intel platforms and with the VM. Allways glitches on the Ryzen 5 (different CPUs and two differens mainboard manufacturers tested)

**Tried solutions:**
- Using mir-kiosk-chromium instead of wpe-webit-mir-kiosk => same problem
- Installing the latest stable MESA driver (from ppa:kisak/kisak-mesa) => same problem
- Installing the latest stable Radeon driver (from https://repo.radeon.com/amdgpu/6.0.2/ubuntu jammy main) => same problem

**Comment by Saviq** [comment]
@mHahnle could you please try on another compositor? E.g. just run `wpe-webkit-mir-kiosk.cog` on GNOME?

---

> Installing the latest stable MESA driver (from ppa:kisak/kisak-mesa) => same problem

FWIW this won't affect snaps, those use a version of https://mir-server.io/docs/the-graphics-core22-snap-interface.

**Comment by AlanGriffiths** [comment]
> could you please try on another compositor

Based on my experience (with #3528) I suspect this is Mir specific, but would be good to confirm

**Comment by mHahnle** [comment]
> @mHahnle could you please try on another compositor? E.g. just run `wpe-webkit-mir-kiosk.cog` on GNOME?

I installed frame-it and started it with:
`frame-it snap run wpe-webkit-mir-kiosk.cog`

Is that what you meant? 

**Result:** no glitches

**Comment by mHahnle** [comment]
> No, just no Mir/Frame in the picture at all. Just run `wpe-webkit-mir-kiosk.cog` in a terminal directly on GNOME.

This doesn't give me any visual output. The terminal just shows: "Setting up watches. Watches established."

**Comment by AlanGriffiths** [comment]
> This doesn't give me any visual output. The terminal just shows: "Setting up watches. Watches established."

Are you able to run GNOME as a Wayland server?

**Comment by AlanGriffiths** [comment]
> That's the output of `inxi -Gaz`:

I mean that "Setting up watches. Watches established." indicates waiting for a Wayland compositor.

If you start GNOME in Wayland mode, then you should be able to run `wpe-webkit-mir-kiosk.cog` on it

**Comment by Saviq** [comment]
> That's why I posted the inxi output. Is it correct? It says X.org as the server, but the composer seems to be gnome?

GNOME can work both on X and Wayland, you need to choose the right session when logging in. WPE is only set up to work on Wayland.

**Comment by mHahnle** [comment]
Yes, I have the login options "Ubuntu" (I guess thats Wayland) and "Ubuntu on Xorg" (I guess thats X). The shown inxi output is from the "Ubuntu" option. The watches messages occure on both.

**Comment by mattkae** [comment]
Out of curiosity, if you set the env variable `mesa_glthread=false`, does that fix it? I was running into something similar earlier on Ubuntu 22.04

**Comment by mHahnle** [comment]
> Out of curiosity, if you set the env variable `mesa_glthread=false`, does that fix it? I was running into something similar earlier on Ubuntu 22.04

IT WORKS!!! Thank you!

Should I close the issue, or is it something that should be integrated by default?

**Comment by AlanGriffiths** [comment]
> Should I close the issue, or is it something that should be integrated by default?

No, the issue remains and this is only a workaround for the underlying problem. But it does provide a clue to help diagnose the underlying problem

**Comment by mattkae** [comment]
> > Should I close the issue, or is it something that should be integrated by default?
> 
> No, the issue remains and this is only a workaround for the underlying problem. But it does provide a clue to help diagnose the underlying problem

Is the underlying problem the fact that this is fixed in mesa in 24.04? Or is the underlying problem in Mir?

**Comment by AlanGriffiths** [comment]
> Is the underlying problem the fact that this is fixed in mesa in 24.04? Or is the underlying problem in Mir?

If there's a fix in more recent Mesa, then I assume switching to the 24/stable of Frame and/or wpe-webkit-mir-kiosk would also work? I think we could close the issue if that were confirmed

**Comment by mHahnle** [comment]
> > Is the underlying problem the fact that this is fixed in mesa in 24.04? Or is the underlying problem in Mir?
> 
> If there's a fix in more recent Mesa, then I assume switching to the 24/stable of Frame and/or wpe-webkit-mir-kiosk would also work? I think we could close the issue if that were confirmed

I did a short test it with Ubuntu Server 24.04 and the 24 channel of Ubuntu Frame and the problem seems gone. (Ubuntu 24 isn't an option for me because of incompatibilities with some used dependecies. But I can live with the workaround, or does disabeling the gltthread has any disadvantages?)

**Comment by AlanGriffiths** [comment]
> I did a short test it with Ubuntu Server 24.04 and the 24 channel of Ubuntu Frame and the problem seems gone. (Ubuntu 24 isn't an option for me because of incompatibilities with some used dependecies. But I can live with the workaround, or does disabeling the gltthread has any disadvantages?)

If the relevant Mesa change is in the userspace drivers (which seems likely), then there should be no need for Ubuntu Server 24.04. So you could switch Frame to 24.

In theory disabling glthread will have some impact on latency, whether that matters for you in practice we can't advise.
